Add script telemetry check when serializing some nodes

pooneh-nb opened the pull request:
#### be889217ea7b2afe7f8c804d929259b9d9fa1008
<pre>
Add script telemetry check when serializing some nodes
<a href="https://bugs.webkit.org/show_bug.cgi?id=283579">https://bugs.webkit.org/show_bug.cgi?id=283579</a>
Include a Radar link (OOPS!).

Reviewed by NOBODY (OOPS!).

Add script telemetry check when serializing values of HTMLInputElement, HTMLSelectElement, and HTMLTextAreaElement.

* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::shouldExcludeElement):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be889217ea7b2afe7f8c804d929259b9d9fa1008

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81370 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/894 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35313 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85899 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32448 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83480 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/912 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8713 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63513 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21326 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84439 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74053 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43808 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28217 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30814 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28801 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87333 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8600 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8781 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69876 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71059 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15116 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14026 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8562 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14085 "Found 1 new failure in editing/MarkupAccumulator.cpp") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8398 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11919 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10206 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -876,8 +879,12 @@ static bool isElementExcludedByRule(const MarkupExclusionRule& rule, const Eleme
+bool MarkupAccumulator::shouldExcludeElement(const Element& element, Node& targetNode)
 {
+    if (targetNode.document().requiresScriptExecutionTelemetry(ScriptTelemetryCategory::FormControls)) {
+        if (is<HTMLInputElement>(element) || is<HTMLSelectElement>(element) || is<HTMLTextAreaElement>(element))
```
Why do we want to do this?

rniwa commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -876,8 +879,12 @@ static bool isElementExcludedByRule(const MarkupExclusionRule& rule, const Eleme
-bool MarkupAccumulator::shouldExcludeElement(const Element& element)
+bool MarkupAccumulator::shouldExcludeElement(const Element& element, Node& targetNode)
 {
+    if (targetNode.document().requiresScriptExecutionTelemetry(ScriptTelemetryCategory::FormControls)) {
```
Why do we need targetNode? Surely, element.document() and targetNode.document() are the same document?

pooneh-nb replied:
That's true, I modified this line.

pooneh-nb committed:
Add script telemetry check when serializing some nodes
https://bugs.webkit.org/show_bug.cgi?id=283579
Include a Radar link (OOPS!).

Reviewed by NOBODY (OOPS!).

Add script telemetry check when serializing values of HTMLInputElement, HTMLSelectElement, and HTMLTextAreaElement.

* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::shouldExcludeElement):

rniwa commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -876,6 +879,10 @@ static bool isElementExcludedByRule(const MarkupExclusionRule& rule, const Eleme
 
 bool MarkupAccumulator::shouldExcludeElement(const Element& element)
 {
+    if (element.document().requiresScriptExecutionTelemetry(ScriptTelemetryCategory::FormControls)) {
```
Safer C++ is complaining that we're calling this non-trivial function (requiresScriptExecutionTelemetry) without storing document in a Ref/RefPtr.
I don't think we can call this function more than once per execution.
We should cache the result of this function call somewhere so that we don't keep repeatedly calling it.

rniwa reviewed (changes requested):
We definitely need to avoid keep calling requiresScriptExecutionTelemetry multiple times since this code is performance sensitive.
